File: artifacts/tulipindicators/translate/src/atr.rs
use crate::*;
use std::assert;
use std::f64;
pub fn ti_atr_start(options: Option<&[f64]>) -> Option<i32> {
    options.map(|opts| opts[0] as i32 - 1)
}
pub fn helper_ti_atr_1(
    sum_ref: &mut f64,
    truerange_ref: &mut f64,
    high: &[f64],
    low: &[f64],
    close: &[f64],
    i: usize,
) {
    let mut sum = *sum_ref;
    let mut truerange = *truerange_ref;

    {
        let l = low[i];
        let h = high[i];
        let c = close[i - 1];
        let ych = (h - c).abs();
        let ycl = (l - c).abs();
        let mut v = h - l;

        if ych > v {
            v = ych;
        }
        if ycl > v {
            v = ycl;
        }
        truerange = v;
    }

    sum += truerange;
    *sum_ref = sum;
    *truerange_ref = truerange;
}
pub fn helper_ti_atr_2(
    output_idx_ref: &mut u32,
    truerange_ref: &mut f64,
    val_ref: &mut f64,
    high: &[f64],
    low: &[f64],
    close: &[f64],
    output: &mut [f64],
    per: f64,
    i: usize,
) {
    let output_idx = *output_idx_ref;
    let mut truerange = *truerange_ref;
    let mut val = *val_ref;

    {
        let l = low[i];
        let h = high[i];
        let c = close[i - 1];
        let ych = (h - c).abs();
        let ycl = (l - c).abs();
        let mut v = h - l;
        
        if ych > v {
            v = ych;
        }
        if ycl > v {
            v = ycl;
        }
        truerange = v;
    }

    val = ((truerange - val) * per) + val;
    output[output_idx as usize] = val;
    
    *output_idx_ref = output_idx + 1;
    *truerange_ref = truerange;
    *val_ref = val;
}
pub fn ti_atr_stream_free(stream: Option<Box<TiStream>>) {
    // In Rust, Box<T> automatically frees memory when it goes out of scope,
    // so we don't need to explicitly free it. Just let the Option<Box> drop.
    // The function takes ownership of the Box and will drop it when done.
}

pub fn helper_ti_atr_stream_run_1(
    output_idx_ref: &mut u32,
    i_ref: &mut i32,
    size: i32,
    stream: &mut TiStreamAtr,
    high: &[f64],
    low: &[f64],
    close: &[f64],
    output: &mut [f64],
    per: f64,
) {
    let mut output_idx = *output_idx_ref;
    let mut i = *i_ref;
    let mut val = stream.last;

    while i < size {
        let truerange = {
            let ych = f64::abs(high[i as usize] - stream.last_close);
            let ycl = f64::abs(low[i as usize] - stream.last_close);
            let mut v = high[i as usize] - low[i as usize];
            if ych > v {
                v = ych;
            }
            if ycl > v {
                v = ycl;
            }
            v
        };

        val = ((truerange - val) * per) + val;
        output[output_idx as usize] = val;
        output_idx += 1;
        stream.last_close = close[i as usize];
        stream.progress += 1;
        i += 1;
    }

    stream.last = val;
    *output_idx_ref = output_idx;
    *i_ref = i;
}
pub fn helper_ti_atr_stream_run_2(
    output_idx_ref: &mut u32,
    i_ref: &mut i32,
    size: i32,
    stream: &mut TiStreamAtr,
    high: &[f64],
    low: &[f64],
    close: &[f64],
    output: &mut [f64],
    per: f64,
    start: i32,
) {
    let mut output_idx = *output_idx_ref;
    let mut i = *i_ref;

    if stream.progress == start {
        stream.sum = high[0] - low[0];
        stream.last_close = close[0];
        stream.progress += 1;
        i += 1;
    }

    while (stream.progress <= 0) && (i < size) {
        let truerange = {
            let ych = (high[i as usize] - stream.last_close).abs();
            let ycl = (low[i as usize] - stream.last_close).abs();
            let mut v = high[i as usize] - low[i as usize];
            if ych > v {
                v = ych;
            }
            if ycl > v {
                v = ycl;
            }
            v
        };

        stream.sum += truerange;
        stream.last_close = close[i as usize];
        stream.progress += 1;
        i += 1;
    }

    if stream.progress == 1 {
        let val = stream.sum * per;
        stream.last = val;
        output[output_idx as usize] = val;
        output_idx += 1;
    }

    *output_idx_ref = output_idx;
    *i_ref = i;
}
pub fn ti_atr(size: usize, inputs: &[&[f64]], options: &[f64], outputs: &mut [&mut [f64]]) -> i32 {
    let high = inputs[0];
    let low = inputs[1];
    let close = inputs[2];
    let period = options[0] as i32;
    let output = &mut outputs[0];
    let mut output_idx: u32 = 0;

    if period < 1 {
        return 1;
    }

    let start = match ti_atr_start(Some(options)) {
        Some(s) => s,
        None => return 1,
    };

    if size <= start as usize {
        return 0;
    }

    let per = 1.0 / period as f64;
    let mut sum = 0.0;
    let mut truerange = 0.0;

    sum += high[0] - low[0];

    for i in 1..period as usize {
        helper_ti_atr_1(&mut sum, &mut truerange, high, low, close, i);
    }

    let mut val = sum / period as f64;
    output[output_idx as usize] = val;
    output_idx += 1;

    for i in period as usize..size {
        helper_ti_atr_2(
            &mut output_idx,
            &mut truerange,
            &mut val,
            high,
            low,
            close,
            output,
            per,
            i,
        );
    }

    assert!((output.as_ptr().wrapping_offset(output_idx as isize) as usize - output.as_ptr() as usize)
        / std::mem::size_of::<f64>()
        == (size - start as usize));
    0
}
pub fn ti_atr_stream_new(options: Option<&[f64]>, stream_in: Option<&mut Option<Box<TiStream>>>) -> Option<i32> {
    // Check if options is None or empty
    let options = options?;
    if options.is_empty() {
        return Some(1);
    }

    let period = options[0] as i32;
    if period < 1 {
        return Some(1);
    }

    // Check if stream_in is None
    let stream_in = stream_in?;

    // Create new TiStreamAtr
    let mut stream = Box::new(TiStreamAtr {
        index: 13,
        progress: -ti_atr_start(Some(options))?,
        period,
        sum: 0.0,
        last: 0.0,
        last_close: 0.0,
    });

    // Convert Box<TiStreamAtr> to Box<TiStream> using unsafe cast
    // This matches the C code's behavior of casting ti_stream_atr* to ti_stream*
    let stream_as_ti: Box<TiStream> = unsafe {
        Box::from_raw(Box::into_raw(stream) as *mut TiStream)
    };

    // Replace the content of stream_in with our new stream
    *stream_in = Some(stream_as_ti);

    Some(0)
}
pub fn ti_atr_stream_run(
    stream_in: &mut TiStream,
    size: i32,
    inputs: &[&[f64]],
    outputs: &mut [&mut [f64]],
) -> i32 {
    let stream = unsafe { &mut *(stream_in as *mut TiStream as *mut TiStreamAtr) };
    let high = inputs[0];
    let low = inputs[1];
    let close = inputs[2];
    let output = &mut outputs[0];
    let mut output_idx = 0;
    let per = 1.0 / (stream.period as f64);
    let start = -(stream.period - 1);
    let mut i = 0;

    if stream.progress < 1 {
        helper_ti_atr_stream_run_2(
            &mut output_idx,
            &mut i,
            size,
            stream,
            high,
            low,
            close,
            output,
            per,
            start,
        );
    }
    if stream.progress >= 1 {
        helper_ti_atr_stream_run_1(
            &mut output_idx,
            &mut i,
            size,
            stream,
            high,
            low,
            close,
            output,
            per,
        );
    }
    0
}
pub fn ti_atr_ref(
    size: i32,
    inputs: Option<&[&[f64]]>,
    options: Option<&[f64]>,
    mut outputs: Option<&mut [&mut [f64]]>,
) -> i32 {
    let period = options.and_then(|opts| opts.first().copied()).map(|p| p as i32);
    if period.map_or(true, |p| p < 1) {
        return 1;
    }

    if size <= ti_atr_start(options).unwrap_or(0) {
        return 0;
    }

    let tr_start = ti_tr_start(options);
    let tr_size = size - tr_start;

    // Use Vec instead of manual malloc/free
    let mut truerange = vec![0.0; tr_size as usize];
    
    // Create a mutable slice reference for tr_outputs
    let tr_outputs = &mut [truerange.as_mut_slice()];
    
    let tr_ret = ti_tr(size as usize, inputs.unwrap_or(&[]), &[], tr_outputs);
    if tr_ret != 0 {
        return tr_ret;
    }

    let wilders_inputs = &[truerange.as_slice()];
    let wilders_ret = ti_wilders(tr_size, Some(wilders_inputs), options, outputs);

    // Assertion translated to debug_assert!
    debug_assert!(
        (size - ti_atr_start(options).unwrap_or(0)) == (size - ti_wilders_start(options).unwrap_or(0))
    );

    wilders_ret.unwrap_or(0)
}

File: artifacts/tulipindicators/decompose/src/atr.c
int ti_atr_start(const double *options)
{
  return ((int) options[0]) - 1;
}


/*** DEPENDENCIES:
***/


int ti_atr(int size, const double * const *inputs, const double *options, double * const *outputs)
{
  const double *high = inputs[0];
  const double *low = inputs[1];
  const double *close = inputs[2];
  const int period = (int) options[0];
  double *output = outputs[0];
  unsigned int output_idx = 0;
  if (period < 1)
  {
    return 1;
  }
  if (size <= ti_atr_start(options))
  {
    return 0;
  }
  const double per = 1.0 / ((double) period);
  double sum = 0;
  double truerange;
  sum += high[0] - low[0];
  int i;
  for (i = 1; i < period; i += 1)
  {
    do
    {
      const double l = low[i];
      const double h = high[i];
      const double c = close[i - 1];
      const double ych = fabs(h - c);
      const double ycl = fabs(l - c);
      double v = h - l;
      if (ych > v)
      {
        v = ych;
      }
      if (ycl > v)
      {
        v = ycl;
      }
      truerange = v;
    }
    while (0);
    sum += truerange;
  }

  double val = sum / period;
  output[output_idx] = val;
  output_idx += 1;
  for (i = period; i < size; i += 1)
  {
    do
    {
      const double l = low[i];
      const double h = high[i];
      const double c = close[i - 1];
      const double ych = fabs(h - c);
      const double ycl = fabs(l - c);
      double v = h - l;
      if (ych > v)
      {
        v = ych;
      }
      if (ycl > v)
      {
        v = ycl;
      }
      truerange = v;
    }
    while (0);
    val = ((truerange - val) * per) + val;
    output[output_idx] = val;
    output_idx += 1;
  }

  assert(((&output[output_idx]) - outputs[0]) == (size - ti_atr_start(options)));
  return 0;
}


/*** DEPENDENCIES:
int ti_atr_start(const double *options)
{
  return ((int) options[0]) - 1;
}


----------------------------
***/


void ti_atr_stream_free(ti_stream *stream)
{
  free(stream);
}


/*** DEPENDENCIES:
struct ti_stream
{
  int index;
  int progress;
}
----------------------------
***/


// hint:  ['output_idx_ref is a mutable refrence to unsigned int']
void helper_ti_atr_stream_run_1(unsigned int * const output_idx_ref, int size, ti_stream_atr * const stream, const double * const high, const double * const low, const double * const close, double * const output, const double per, int i)
{
  unsigned int output_idx = *output_idx_ref;
  double val = stream->last;
  while (i < size)
  {
    double truerange;
    do
    {
      const double ych = fabs(high[i] - stream->last_close);
      const double ycl = fabs(low[i] - stream->last_close);
      double v = high[i] - low[i];
      if (ych > v)
      {
        v = ych;
      }
      if (ycl > v)
      {
        v = ycl;
      }
      truerange = v;
    }
    while (0);
    val = ((truerange - val) * per) + val;
    output[output_idx] = val;
    output_idx += 1;
    stream->last_close = close[i];
    stream->progress += 1;
    ++i;
  }

  stream->last = val;
  *output_idx_ref = output_idx;
}


/*** DEPENDENCIES:
struct ti_stream
{
  int index;
  int progress;
}
----------------------------
typedef struct ti_stream_atr
{
  int index;
  int progress;
  int period;
  double sum;
  double last;
  double last_close;
} ti_stream_atr
----------------------------
***/


// hint:  ['output_idx_ref is a mutable refrence to unsigned int', 'i_ref is a mutable refrence to int']
void helper_ti_atr_stream_run_2(unsigned int * const output_idx_ref, int * const i_ref, int size, ti_stream_atr * const stream, const double * const high, const double * const low, const double * const close, double * const output, const double per, const int start)
{
  unsigned int output_idx = *output_idx_ref;
  int i = *i_ref;
  if (stream->progress == start)
  {
    stream->sum = high[0] - low[0];
    stream->last_close = close[0];
    stream->progress += 1;
    i += 1;
  }
  while ((stream->progress <= 0) && (i < size))
  {
    double truerange;
    do
    {
      const double ych = fabs(high[i] - stream->last_close);
      const double ycl = fabs(low[i] - stream->last_close);
      double v = high[i] - low[i];
      if (ych > v)
      {
        v = ych;
      }
      if (ycl > v)
      {
        v = ycl;
      }
      truerange = v;
    }
    while (0);
    stream->sum += truerange;
    stream->last_close = close[i];
    stream->progress += 1;
    i += 1;
  }

  if (stream->progress == 1)
  {
    const double val = stream->sum * per;
    stream->last = val;
    output[output_idx] = val;
    output_idx += 1;
  }
  *output_idx_ref = output_idx;
  *i_ref = i;
}


/*** DEPENDENCIES:
struct ti_stream
{
  int index;
  int progress;
}
----------------------------
typedef struct ti_stream_atr
{
  int index;
  int progress;
  int period;
  double sum;
  double last;
  double last_close;
} ti_stream_atr
----------------------------
***/


int ti_atr_stream_new(const double *options, ti_stream **stream_in)
{
  ti_stream_atr **stream = (ti_stream_atr **) stream_in;
  const int period = (int) options[0];
  if (period < 1)
  {
    return 1;
  }
  *stream = malloc(sizeof(ti_stream_atr));
  if (!(*stream))
  {
    return 2;
  }
  (*stream)->index = 13;
  (*stream)->progress = -ti_atr_start(options);
  (*stream)->period = period;
  (*stream)->sum = 0.0;
  return 0;
}


/*** DEPENDENCIES:
int ti_atr_start(const double *options)
{
  return ((int) options[0]) - 1;
}


----------------------------
struct ti_stream
{
  int index;
  int progress;
}
----------------------------
typedef struct ti_stream_atr
{
  int index;
  int progress;
  int period;
  double sum;
  double last;
  double last_close;
} ti_stream_atr
----------------------------
***/


int ti_atr_stream_run(ti_stream *stream_in, int size, const double * const *inputs, double * const *outputs)
{
  ti_stream_atr *stream = (ti_stream_atr *) stream_in;
  const double *high = inputs[0];
  const double *low = inputs[1];
  const double *close = inputs[2];
  double *output = outputs[0];
  unsigned int output_idx = 0;
  const double per = 1.0 / ((double) stream->period);
  const int start = -(stream->period - 1);
  int i = 0;
  if (stream->progress < 1)
  {
    helper_ti_atr_stream_run_2(&output_idx, &i, size, stream, high, low, close, output, per, start);
  }
  if (stream->progress >= 1)
  {
    helper_ti_atr_stream_run_1(&output_idx, size, stream, high, low, close, output, per, i);
  }
  return 0;
}


/*** DEPENDENCIES:
void helper_ti_atr_stream_run_1(unsigned int * const output_idx_ref, int size, ti_stream_atr * const stream, const double * const high, const double * const low, const double * const close, double * const output, const double per, int i)
{
  unsigned int output_idx = *output_idx_ref;
  double val = stream->last;
  while (i < size)
  {
    double truerange;
    do
    {
      const double ych = fabs(high[i] - stream->last_close);
      const double ycl = fabs(low[i] - stream->last_close);
      double v = high[i] - low[i];
      if (ych > v)
      {
        v = ych;
      }
      if (ycl > v)
      {
        v = ycl;
      }
      truerange = v;
    }
    while (0);
    val = ((truerange - val) * per) + val;
    output[output_idx] = val;
    output_idx += 1;
    stream->last_close = close[i];
    stream->progress += 1;
    ++i;
  }

  stream->last = val;
  *output_idx_ref = output_idx;
}


----------------------------
typedef struct ti_stream_atr
{
  int index;
  int progress;
  int period;
  double sum;
  double last;
  double last_close;
} ti_stream_atr
----------------------------
void helper_ti_atr_stream_run_2(unsigned int * const output_idx_ref, int * const i_ref, int size, ti_stream_atr * const stream, const double * const high, const double * const low, const double * const close, double * const output, const double per, const int start)
{
  unsigned int output_idx = *output_idx_ref;
  int i = *i_ref;
  if (stream->progress == start)
  {
    stream->sum = high[0] - low[0];
    stream->last_close = close[0];
    stream->progress += 1;
    i += 1;
  }
  while ((stream->progress <= 0) && (i < size))
  {
    double truerange;
    do
    {
      const double ych = fabs(high[i] - stream->last_close);
      const double ycl = fabs(low[i] - stream->last_close);
      double v = high[i] - low[i];
      if (ych > v)
      {
        v = ych;
      }
      if (ycl > v)
      {
        v = ycl;
      }
      truerange = v;
    }
    while (0);
    stream->sum += truerange;
    stream->last_close = close[i];
    stream->progress += 1;
    i += 1;
  }

  if (stream->progress == 1)
  {
    const double val = stream->sum * per;
    stream->last = val;
    output[output_idx] = val;
    output_idx += 1;
  }
  *output_idx_ref = output_idx;
  *i_ref = i;
}


----------------------------
struct ti_stream
{
  int index;
  int progress;
}
----------------------------
***/


int ti_atr_ref(int size, const double * const *inputs, const double *options, double * const *outputs)
{
  const int period = (int) options[0];
  if (period < 1)
  {
    return 1;
  }
  if (size <= ti_atr_start(options))
  {
    return 0;
  }
  const int tr_start = ti_tr_start(options);
  const int tr_size = size - tr_start;
  double *truerange = malloc(((unsigned int) tr_size) * (sizeof(double)));
  if (!truerange)
  {
    return 2;
  }
  double *tr_outputs[1] = {truerange};
  const int tr_ret = ti_tr(size, inputs, 0, tr_outputs);
  if (tr_ret != 0)
  {
    free(truerange);
    return tr_ret;
  }
  const double *wilders_inputs[1] = {truerange};
  const int wilders_ret = ti_wilders(tr_size, wilders_inputs, options, outputs);
  free(truerange);
  assert((size - ti_atr_start(options)) == (size - ti_wilders_start(options)));
  return wilders_ret;
}


/*** DEPENDENCIES:
int ti_wilders(int size, const double * const *inputs, const double *options, double * const *outputs)
{
  const double *input = inputs[0];
  const int period = (int) options[0];
  double *output = outputs[0];
  unsigned int output_idx = 0;
  if (period < 1)
  {
    return 1;
  }
  if (size <= ti_wilders_start(options))
  {
    return 0;
  }
  const double per = 1.0 / ((double) period);
  double sum = 0;
  int i;
  for (i = 0; i < period; i += 1)
  {
    sum += input[i];
  }

  double val = sum / period;
  output[output_idx] = val;
  output_idx += 1;
  for (i = period; i < size; i += 1)
  {
    val = ((input[i] - val) * per) + val;
    output[output_idx] = val;
    output_idx += 1;
  }

  assert(((&output[output_idx]) - outputs[0]) == (size - ti_wilders_start(options)));
  return 0;
}


----------------------------
int ti_atr_start(const double *options)
{
  return ((int) options[0]) - 1;
}


----------------------------
int ti_tr(int size, const double * const *inputs, const double *options, double * const *outputs)
{
  const double *high = inputs[0];
  const double *low = inputs[1];
  const double *close = inputs[2];
  (void) options;
  double *output = outputs[0];
  double truerange;
  output[0] = high[0] - low[0];
  int i;
  for (i = 1; i < size; i += 1)
  {
    do
    {
      const double l = low[i];
      const double h = high[i];
      const double c = close[i - 1];
      const double ych = fabs(h - c);
      const double ycl = fabs(l - c);
      double v = h - l;
      if (ych > v)
      {
        v = ych;
      }
      if (ycl > v)
      {
        v = ycl;
      }
      truerange = v;
    }
    while (0);
    output[i] = truerange;
  }

  return 0;
}


----------------------------
int ti_wilders_start(const double *options)
{
  return ((int) options[0]) - 1;
}


----------------------------
int ti_tr_start(const double *options)
{
  (void) options;
  return 0;
}


----------------------------
***/


